nixos/lib/test-driver/test-driver.py

```diff
@@ -0,0 +1,685 @@
+                status, _ = self.execute('[ -e /tmp/.X11-unix/X0 ]')
+                if status == 0:
+                    return
+
+    def sleep(self, secs):
+        time.sleep(secs)
+
+    def block(self):
+        """Make the machine unreachable by shutting down eth1 (the multicast
+        interface used to talk to the other VMs).  We keep eth0 up so that
+        the test driver can continue to talk to the machine.
+        """
+        self.send_monitor_command('set_link virtio-net-pci.1 off')
+
+    def unblock(self):
+        """Make the machine reachable.
+        """
+        self.send_monitor_command('set_link virtio-net-pci.1 on')
+
+
+log = Logger()
+
+
+def create_machine(args):
+    global log
+    args['log'] = log
+    args['redirectSerial'] = os.environ.get('USE_SERIAL', '0') == '1'
+    return Machine(args)
+
+
+vlan_nrs = list(dict.fromkeys(os.environ['VLANS'].split()))
+
+vde_sockets = [create_vlan(v) for v in vlan_nrs]
+
+for nr, vde_socket, _, _ in vde_sockets:
+    os.environ['QEMU_VDE_SOCKET_{}'.format(nr)] = vde_socket
+
+vm_scripts = sys.argv[1:]
+
+machines = [create_machine({'startCommand': s}) for s in vm_scripts]
+machine_eval = ['{0} = machines[{1}]'.format(m.name, idx)
+                for idx, m in enumerate(machines)]
+exec('\n'.join(machine_eval))
+
+
+def start_all():
+    with log.nested('starting all VMs'):
+        for machine in machines:
+            machine.start()
+
+
+def join_all():
+    with log.nested('waiting for all VMs to finish'):
+        for machine in machines:
+            machine.wait_for_shutdown()
+
+
+def run_tests():
+    with log.nested('running the VM test script'):
+        test_script = os.environ['testScript']
+        try:
+            exec(test_script)
+        except Exception as e:
+            eprint('error: {}'.format(str(e)))
+            sys.exit(1)
+
+    # TODO: Collect coverage data
+
+    for machine in machines:
+        if machine.is_up():
+            machine.execute('sync')
+
+    if nr_tests != 0:
+        log.log('{} out of {} tests succeeded'.format(nr_succeeded, nr_tests))
+
+
+class subtest:
+    def __init__(self, name):
+        self.name = name
+
+    def __enter__(self):
+        global nr_tests
+        self.tic = log.nest_enter(self.name)
+        nr_tests += 1
+
+        return
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        global nr_succeeded
+        if exc_type is None:
+            nr_succeeded += 1
+        else:
+            log.log('error: {}'.format(str(exc_value)))
+
+        log.nest_exit(self.tic)
+
+        if exc_type is not None:
+            return True
+
+
+nr_tests = 0
+nr_succeeded = 0
+
+
+@atexit.register
+def clean_up():
+    with log.nested('cleaning up'):
+        for machine in machines:
+            if machine.pid is None:
+                continue
+            log.log('killing {} (pid {})'.format(machine.name, machine.pid))
+            machine.process.kill()
+
+    log.close()
+
+
+tic = time.time()
```

Comment:
```python
if __name__ == "__main__":
  begin = time.time()
  run_tests()
  end = time.time()
  print('test script finished in {:.2f}s'.format(end - begin))
```

Reply:
why would you only move those 4 lines into the conditional, while the rest of the script already started all the machines? what is the use case?

Reply:
Actually I would put it in a `main()` function, so that later one something like mypy can be used for the driver for type-checking.

Reply:
I second that, being able to use this file as a library has many benefits (including easier understanding when reading the code).

Also gets rid of global state for when [strict python](https://instagram-engineering.com/python-at-scale-strict-modules-c0bb9245c834) becomes a thing :)